Autotest that checks login function for user that has registered e-mail and password

MPolishchuk8 opened the pull request:
In this autotest i realized login function with updates regarding comments given.

MPolishchuk8 committed:
Add files via upload

e-genov-ks commented:
> Це навіть не 5 хвилин, чекаю коментарі)

в описі пул реквесту люди додають опис змін, щоб легше було розуміти, що саме в коді. Ти фактично місце з чатом сплутав )

e-genov-ks commented on README.md:
чому цей файл потрапив в ПР ?

MPolishchuk8 replied:
Вніс зміни

e-genov-ks commented on src/main/java/org/example/Main.java:
```diff
@@ -0,0 +1,7 @@
+package org.example;
+
+public class Main {
+    public static void main(String[] args) {
+        System.out.println("Hello world!");
+    }
+}
```
для чого цей файл ? Яка його відповідальність?

MPolishchuk8 replied:
Видалив

e-genov-ks commented on pom.xml:
```diff
@@ -0,0 +1,47 @@
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
+
+</project>
```
додавай завжди newline, про це і гітхаб каже

MPolishchuk8 replied:
Додав

e-genov-ks commented on src/test/java/org/example/BaseTest.java:
```diff
@@ -0,0 +1,59 @@
+
+public class BaseTest {
+    public WebDriver driver;
+    protected FrontPage frontPage;
+    protected LoginPage loginPage;
+    protected ProfilePage profilePage;
```
в чому сенс тримати ці речі в baseTest ? BaseTest - абстракція, а фронт/логін/профайл сторінки - це реалізація певна.

MPolishchuk8 replied:
Переніс в LoginTest

e-genov-ks commented on src/test/java/org/example/BaseTest.java:
```diff
@@ -0,0 +1,59 @@
+
+    @BeforeEach
+    public void setUp() {
+        System.setProperty("webdriver.chrome.driver", "C:\\chromedriver\\chrome\\chromedriver.exe");
```
в мене мак. як мені запустити цей код ? )

MPolishchuk8 replied:
Використав WebDriverManager

e-genov-ks commented on src/test/java/org/example/BaseTest.java:
```diff
@@ -0,0 +1,59 @@
+        options.addArguments("--start-maximized");
+        driver = new ChromeDriver(options);
+        driver.get(ConfProperties.getProperty("loginpage"));
+        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
```
в чому сенс цього wait ? для чого чекати аж 10 секунд? уяви, що в тебе 1000 тестів і кожен буде чекати по 10 секунд,...

MPolishchuk8 replied:
Прописав загальний таймаут Configuration.timeout = 2000 , старий вейт прибрав

e-genov-ks commented on src/test/java/org/example/ConfProperties.java:
```diff
@@ -0,0 +1,32 @@
+    protected static FileInputStream fileInputStream;
+    protected static Properties PROPERTIES;
+
+    static {
+        try {
+            fileInputStream = new FileInputStream("src/test/resources/conf.properties");
+            PROPERTIES = new Properties();
+            PROPERTIES.load(fileInputStream);
+        } catch (IOException e) {
+            e.printStackTrace();
+
+        } finally {
+            if (fileInputStream != null)
+                try {
+                    fileInputStream.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+        }
+    }
```
в когось просто скопіпастив цей код. Ну ок, проговоримо його голосом ))))

MPolishchuk8 replied:
Усі properties беруться з Environment variables

e-genov-ks commented on src/test/java/org/example/LoginTest.java:
```diff
@@ -0,0 +1,21 @@
+    void loginTest() {
+        frontPage.clickLoginBtn();
+        switchToNewWindow();
+        driver.manage().window().maximize();
```
чого тільки тут на весь екран? Чому попередні кроку без цього?

e-genov-ks commented on src/test/java/org/example/LoginTest.java:
```diff
@@ -0,0 +1,21 @@
+        loginPage.inputLogin(ConfProperties.getProperty("login"));
+        loginPage.clickLoginBtn();
+        loginPage.inputPasswd(ConfProperties.getProperty("password"));
+        loginPage.clickLoginBtn();
```
для чого двічі натискати кнопку логіна ?

MPolishchuk8 replied:
Змінив на NextBtn

e-genov-ks commented on src/test/java/org/example/LoginTest.java:
```diff
@@ -0,0 +1,21 @@
+class LoginTest extends BaseTest {
+
+    @Test
+    void loginTest() {
```
що саме тестує цей тест? по назві не зрозуміло

MPolishchuk8 replied:
Додав коментар з поясненням, можливо ти мав на увазі щось інше

e-genov-ks replied:
Це найгірше, що можна було зробити. Коментарі - це дуже погано :) метод має називатись правильно, а не коментарі дописуватись

e-genov-ks commented on src/test/java/org/example/pages/FrontPage.java:
```diff
@@ -0,0 +1,28 @@
+
+    public FrontPage(WebDriver driver) {
+        this.driver = driver;
+        this.wait = new WebDriverWait(driver, Duration.ofSeconds(10));
```
знову ці 10 секунд, які були вже в basePage

MPolishchuk8 replied:
Прибрав, глобальний таймаут прописаний

e-genov-ks commented on src/test/java/org/example/pages/FrontPage.java:
```diff
@@ -0,0 +1,28 @@
+    public FrontPage(WebDriver driver) {
+        this.driver = driver;
+        this.wait = new WebDriverWait(driver, Duration.ofSeconds(10));
+        PageFactory.initElements(driver, this);
+        PageFactory.initElements(driver, this);
```
просто задублював рядки

MPolishchuk8 replied:
Видалив зайве

e-genov-ks commented on src/test/java/org/example/pages/LoginPage.java:
```diff
@@ -0,0 +1,39 @@
+        PageFactory.initElements(driver, this);
+    }
+
+    @FindBy(xpath = "//*[contains(text(), 'Продовжити')]")
```
а якщо буде англійська мова ?

MPolishchuk8 replied:
Зробив пошук елементу по data-testid

e-genov-ks commented on src/test/java/org/example/pages/ProfilePage.java:
```diff
@@ -0,0 +1,84 @@
+/* package org.example.pages;
```
для чого лишати коментарі ?

MPolishchuk8 replied:
Прибрав

e-genov-ks commented on src/test/java/org/example/pages/ProfilePage.java:
```diff
@@ -0,0 +1,84 @@
+        PageFactory.initElements(driver, this);
+    }
+
+    @FindBy(xpath = "//*[contains(@class, 'UserProfileContainer_subButton__loxWR')]")
```
це погана практика шукати елементи по цсс-класам

MPolishchuk8 replied:
Змінив на пошук по title

e-genov-ks commented on src/test/resources/conf.properties:
```diff
@@ -0,0 +1,3 @@
+loginpage = https://hub.kyivstar.ua/
+login = kshublogintest@gmail.com
+password = Test678test
```
хардкодити паролі в репозиторія - дуже не правильно, безпека за таке з'їсть

MPolishchuk8 replied:
Все занесено в Environment variables

e-genov-ks commented on target/test-classes/org/example/BaseTest.class:
для чого це було комітити ? це артефакти білда, їх не має бути в коді

MPolishchuk8 replied:
змінив

e-genov-ks commented:
ну і назва ПР не відповідає тому що в коді. В назві "Add files via upload"

e-genov-ks commented on pom.xml:
```diff
@@ -0,0 +1,47 @@
+
+    <dependencies>
+        <!-- https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-java -->
+        <dependency>
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>selenium-java</artifactId>
+            <version>4.23.1</version>
+        </dependency>
```
зрозумій для чого ця бібліотека

MPolishchuk8 replied:
Зрозумів

e-genov-ks commented on src/test/java/org/example/BaseTest.java:
```diff
@@ -0,0 +1,59 @@
+
+public class BaseTest {
+    public WebDriver driver;
+    protected FrontPage frontPage;
+    protected LoginPage loginPage;
+    protected ProfilePage profilePage;
```
почитай про Single Responsibility.

MPolishchuk8 replied:
Прочитав і переніс в LoginTest

e-genov-ks commented on src/test/java/org/example/BaseTest.java:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    protected void switchToNewWindow() {
+        String originalWindow = driver.getWindowHandle();
+        Set<String> allWindows = driver.getWindowHandles();
+        for (String windowHandle : allWindows) {
+            if (!windowHandle.equals(originalWindow)) {
+                driver.switchTo().window(windowHandle);
+                break;
+            }
+        }
+    }
```
в даному випадку все буде працювати, тільки коли в тебе 1 віконечко. Якщо їх декілька - буде помилка

MPolishchuk8 replied:
Condition added

MPolishchuk8 replied:
Додав перевірку вікон що відкриваються за ідентифікатором

e-genov-ks commented on src/test/java/org/example/BaseTest.java:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    protected void switchToNewWindow() {
+        String originalWindow = driver.getWindowHandle();
+        Set<String> allWindows = driver.getWindowHandles();
+        for (String windowHandle : allWindows) {
+            if (!windowHandle.equals(originalWindow)) {
+                driver.switchTo().window(windowHandle);
+                break;
+            }
+        }
+    }
```
чого цей код саме тут ? Це BaseTest, а це метод про певну реалізацію

MPolishchuk8 replied:
Moved to MainPage

MPolishchuk8 replied:
Moved to MainPage

MPolishchuk8 replied:
Переніс в MainPage

e-genov-ks commented on src/test/java/org/example/pages/ProfilePage.java:
```diff
@@ -0,0 +1,84 @@
+
+    public ProfilePage(WebDriver driver) {
+        this.driver = driver;
+        this.wait = new WebDriverWait(driver, Duration.ofSeconds(10));
```
для вивчення https://www.selenium.dev/documentation/webdriver/waits/

MPolishchuk8 committed:
Make changes to project after code review

Add Selenide library
Change waits
Delete extra comments
Add proper elements search
Create Environment variables
Add identification to opening windows

MPolishchuk8 committed:
Delete target directory

MPolishchuk8 committed:
Delete README.md

MPolishchuk8 committed:
Delete src/main/java/org/example directory

MPolishchuk8 committed:
Delete src/test/resources directory

MPolishchuk8 committed:
Delete src/test/java/org/example/pages/FrontPage.java

MPolishchuk8 committed:
Delete src/test/java/org/example/ConfProperties.java

e-genov-ks commented on src/test/java/org/example/pages/LoginPage.java:
```diff
@@ -0,0 +1,51 @@
+        Set<String> allWindows = driver.getWindowHandles();
+        for (String windowHandle : allWindows) {
+            if (!windowHandle.equals(originalWindow)) {
+                driver.switchTo().window(windowHandle);
+                if ($(By.xpath("//*[@data-testid='login-input']")).exists()) break;
```
шо це за логіка така ? він перемикається на вікно і потім виходить...

e-genov-ks commented on src/test/java/org/example/pages/MainPage.java:
```diff
@@ -0,0 +1,32 @@
+        $("#login-button").shouldBe(clickable).click();
+    }
+
+    public void switchToNewWindow() {
+        String originalWindow = driver.getWindowHandle();
+        Set<String> allWindows = driver.getWindowHandles();
+        for (String windowHandle : allWindows) {
+            if (!windowHandle.equals(originalWindow)) {
+                driver.switchTo().window(windowHandle);
+                if ($(By.xpath("//*[@data-testid='login-input']")).exists()) break;
+            }
+        }
+    }
```
цей код дублюється, був в LoginPage

e-genov-ks commented on src/test/java/org/example/pages/ProfilePage.java:
```diff
@@ -0,0 +1,39 @@
+        this.driver = driver;
+    }
+
+    private SelenideElement userMenu = $x("//*[contains(@title, 'login')]");
```
краще замість contains використовувати equal , бо contains це не строгий пошук

MPolishchuk8 committed:
Make changes to project after 2nd round of code review

Give correct method name
Change method switchToNewWindow
Make proper element search on profile page

MPolishchuk8 committed:
Merge remote-tracking branch 'origin/autotest' into autotest

Commit:
fix

e-genov-ks commented on src/test/java/org/example/pages/LoginPage.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public void switchToNewWindow() {
+        String originalWindow = driver.getWindowHandle();
```
помилка тут :)

Commit:
fix

Commit:
remove unused imports

Commit:
remove unused imports

MPolishchuk8 committed:
Merge pull request #2 from MPolishchuk8/autotest-fix

Autotest fix

MPolishchuk8 committed:
Make changes to BaseTest and LoginTest

Remove unused import
Add wait to switchToWindow

e-genov-ks commented on src/test/java/org/example/LoginTest.java:
```diff
@@ -0,0 +1,38 @@
+        loginPage.inputPasswd();
+        loginPage.clickNextBtn();
+        switchToWindow(Config.PRODUCT_URL);
+        try {
+            Thread.sleep(1000);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
```
Для чого в сліп інструкції трай-кетч ?

MPolishchuk8 replied:
Воно зроблено для перехоплення виключення переривання потоку під час виконання сліпу

MPolishchuk8 reviewed (commented):
Воно зроблено для перехоплення виключення переривання потоку під час виконання сліпу

e-genov-ks commented on src/test/java/org/example/LoginTest.java:
```diff
@@ -0,0 +1,38 @@
+class LoginTest extends BaseTest {
+
+    @Test
+    void assertLoginSuccessWithCorrectCredentials() {
```
```suggestion
    void assertLoginSuccessWithCorrectCredentials() throws InterruptedException {

```

e-genov-ks commented on src/test/java/org/example/LoginTest.java:
```diff
@@ -0,0 +1,38 @@
+        loginPage.inputPasswd();
+        loginPage.clickNextBtn();
+        switchToWindow(Config.PRODUCT_URL);
+        try {
+            Thread.sleep(1000);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
```
```suggestion
        Thread.sleep(1000);
```

e-genov-ks reviewed (commented):
Запропонував рішення без try/catch, щоб це норм читалось

MPolishchuk8 committed:
Make changes to Thread.sleep

Change Thread.sleep to make code more readable
